### PR TITLE
NodeList wrapper fixed (closes #1111)

### LINKED ISSUE #1111: Performance slow down
Firefox profiler after testcafe qunit test execution
https://www.screencast.com/t/ujSwtpHN - current release version (10.6.8)
https://www.screencast.com/t/u3UMlLIomqow - previous version (10.6.7)

This slow-down appeared due to this commit
https://github.com/DevExpress/testcafe-hammerhead/commit/885436e6090d968e5c44ac7cb2deae9765865be5

### PR DESCRIPTION
**Pull request by georgiy-abbasov**
\cc @churkin, @AlexanderMoskovkin 

Changes:
- split `_filterNodeList` to `_filterLiveNodeList` and `_filterNonLiveNodeList`
See https://github.com/ryanmcdermott/clean-code-javascript#dont-use-flags-as-function-parameters

- NodeListWrapper for live and non-live node collections
- using destructing syntax  for `createNodeListWrapper`
See https://github.com/ryanmcdermott/clean-code-javascript#function-arguments-2-or-fewer-ideally

### CODE REVIEW

File: src/client/sandbox/node/node-list-wrapper.js [hunk 1]
```diff
@@ -15,27 +15,15 @@ export default class NodeListWrapper {
...
-            value: filterItemFn
-        });
+        Object.defineProperty(this, '_nodeList', { value: nodeList });
 
-        Object.defineProperty(this, '_filteredNodeList', {
-            get: () => this._filterListFn(this._nodeList, this._filterItemFn)
-        });
+        Object.defineProperty(this, '_filterListFn', { value: filterListFn });
 
+        Object.defineProperty(this, '_filterItemFn', { value: filterItemFn });
```
Comment (AlexanderMoskovkin) [inline src/client/sandbox/node/node-list-wrapper.js]: I think `_filterItemFn` is redundant here. We can do the following when the wrapper is created:
```js
return new NodeListWrapper({
    nodeList,
    filterListFn: items => ShadowUI._filterList(items, ShadowUI._filterElement),
    isLiveCollection
});
```

File: src/client/sandbox/node/node-list-wrapper.js [hunk 2]
```diff
@@ -44,13 +32,15 @@ export default class NodeListWrapper {
...
+        this._prevNodeListLength = this._nodeList.length;
+
         NodeListWrapper._refreshNodeList.call(this);
     }
 
     static _refreshNodeList () {
         this._filteredNodeList = this._filterListFn(this._nodeList, this._filterItemFn);
 
-        var i;
+        var i = null;
```
Comment (AlexanderMoskovkin) [inline src/client/sandbox/node/node-list-wrapper.js]: initialize it as number if you are going to use it as a number.
```js
var i = 0;

for(; i< this....; i++) {
...
}
```

File: test/client/fixtures/sandbox/shadow-ui-test.js [hunk 3]
```diff
@@ -835,3 +835,34 @@ test('getElementsByTagName', function () {
...
+
+    strictEqual(elements.length, 1);
+    strictEqual(elements[0], textarea1);
+    strictEqual(elements[1], void 0);
+
+    testDiv.appendChild(textarea2);
+    testDiv.appendChild(textarea3);
+    testDiv.appendChild(textarea4);
+
+    strictEqual(elements.length, 4);
```
Comment (AlexanderMoskovkin) [inline test/client/fixtures/sandbox/shadow-ui-test.js]: just comment this assertion and your test is failed. It's not ok.
The same for other tests

File: test/client/fixtures/sandbox/shadow-ui-test.js [hunk 4]
```diff
@@ -835,3 +835,34 @@ test('getElementsByTagName', function () {
...
+    testDiv.appendChild(shadowUIElement);
+
+    var elements = document.getElementsByTagName('textarea');
+
+    strictEqual(elements.length, 1);
+    strictEqual(elements[0], textarea1);
+    strictEqual(elements[1], void 0);
+
+    testDiv.appendChild(textarea2);
+    testDiv.appendChild(textarea3);
```
Comment (AlexanderMoskovkin) [inline test/client/fixtures/sandbox/shadow-ui-test.js]: is there any reason to add more than one additional textarea here? As I see, one is enough

File: test/client/fixtures/sandbox/shadow-ui-test.js (hunk 3 above)
Comment (georgiy-abbasov) [inline test/client/fixtures/sandbox/shadow-ui-test.js]: We decided to refresh collection when getting its length. Most users get length before getting item by index.

Comment (AlexanderMoskovkin) [inline test/client/fixtures/sandbox/shadow-ui-test.js]: > Most users ...

Who are they? Have you accomplished a research about it? =0 And what about others like me, who has a script on the page that doesn't call `items.length` before getting an item by its index?

Anyway it seems changes are required here. I believe you with @churkin will find the best approach

**Comment by georgiy-abbasov** [comment]
Recalculating filtered nodelist on property [getter](https://github.com/DevExpress/testcafe-hammerhead/pull/1112/files#diff-3de170469e255c16d450dfffc8b23e1aR26) gives slowdown too - 
https://www.screencast.com/t/TEcU651HSOJn
vs
https://www.screencast.com/t/fEhpnrGCQ (without recalculating)

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by georgiy-abbasov** [comment]
FPR

**Comment by miherlosev** [comment]
After merge 885436e a lot of sites become to hung in health-monitor.
Let's run health-monitor on this fork/branch before merging.

**Comment by georgiy-abbasov** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot /retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
@testcafe-build-bot \retest

**Comment by miherlosev** [comment]
FPR
@AlexanderMoskovkin @churkin @LavrovArtem 
I've update the list changes in this PR. See [description](https://github.com/DevExpress/testcafe-hammerhead/pull/1112#issue-226030978)

After all reviews were completed I will run HealthMonitor on this PR.

**Comment by AlexanderMoskovkin** [comment]
Have you checked this approach with health-monitor?

**Comment by miherlosev** [comment]
I run health monitor for this PR

**Comment by miherlosev** [comment]
There is the perfomance degradation - https://testcafe-hhhm.devexpress.com/history?timestamp=1495658848246&siteUrl=http://4dsply.com/
I am investigating it...

**Comment by miherlosev** [comment]
difficult to merge. It will be fixed in another PR.

**Pull request closed without merging**